### PR TITLE
[Session] Adding the Session type

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

Scaffolding & adding the `Session` type

### Summary generated by Reviewpad on 11 Oct 23 22:59 UTC
This pull request adds a new file called `session.proto` to the `pocket/session` directory. This file defines the protobuf messages `SessionHeader` and `Session`, which represent lightweight and fully-hydrated session objects respectively. It also imports several other protobuf files and sets options for the Go package. Overall, this pull request adds the necessary code for managing sessions in the Pocket application.

## Issue

Part of #10

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make test_all_unit`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by Olshansk** [comment]
@bryanchriswhite @red-0ne This isn't ready to be merged in, but I think it's close and would really appreciate your feedback on the first iteration of its design.

File: proto/pocket/session/session.proto [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+option go_package = "pocket/x/session/types";
+
+import "cosmos_proto/cosmos.proto";
+import "pocket/service/service.proto";
+import "pocket/application/application.proto";
+import "pocket/supplier/supplier.proto";
+
+// SessionHeader is a lightweight header for a session that can be passed around
+// and from which the full Session object can be hydrated.
+message SessionHeader {
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply proto/pocket/session/session.proto:13]: I understand your reasoning. What do you think of this updated `SessionHeader`

```go
// SessionHeader is a lightweight header for a session that can be passed around.
// It is the minimal amount of data required to hydrate & retrieve all data relevant to the session.
message SessionHeader {
    string application_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The address of the application this session is for
    service.ServiceId service_id = 2; // The ID of the service this session is servicing
    uint64 session_block_start_height = 3; // The height at which this session started
    // NOTE: session_id can be derived from the above values using on-chain but is included in the header for convenience
    string session_id = 4; // A unique pseudoranom ID for this session
}
```

@bryanchriswhite Lmk if this looks good to you and I'll make the changes if so.

[the remark this answers is not in the document]
Comment (Olshansk) [reply proto/pocket/session/session.proto:13]: @red-0ne Lmk what you think before I edit it.

[the remark this answers is not in the document]
Comment (red-0ne) [reply proto/pocket/session/session.proto:13]: @Olshansk , This is great! `session_block_start_height` and maybe `session_block_end_height` are totally viable alternatives.

For `session_id` as long as the on-chain context does not rely on it and gets its own from the above (verified) params then all good :+1:

**Review by red-0ne — DISMISSED** [review]
LGTM :+1:

File: proto/pocket/session/session.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply proto/pocket/session/session.proto:13]: 👌

[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply proto/pocket/session/session.proto:13]: Are we still considering having 2 session header structs or does this conversation mean we only need the one?

File: proto/pocket/session/session.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    string session_id = 1; // A unique pseudoranom ID for this session
+    uint64 session_number = 2; // The session number since genesis
+    uint64 session_block_start_height = 3; // The height at which this session started
```
Comment (red-0ne) [inline proto/pocket/session/session.proto:16]: Should we align with CosmosSDK and have block heights as `int64`s

File: proto/pocket/session/session.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply proto/pocket/session/session.proto:13]: Just one

**Review by bryanchriswhite — APPROVED** [review]
🚀

**Review by red-0ne — APPROVED** [review]
Perfect.

**Comment by bryanchriswhite** [comment]
I'm recalling that when I was resolving the dependency cycle between session and supplier, part of the solution was to move the `Supplier` and/or `SessionHeader` (in this case) to a "shared" module.

**Review by bryanchriswhite — DISMISSED** [review]

File: proto/pocket/session/session.proto [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// NOTE: Using signed integers for consistency with the cosmos SDK
+
+// SessionHeader is a lightweight header for a session that can be passed around.
+// It is the minimal amount of data required to hydrate & retrieve all data relevant to the session.
+message SessionHeader {
+    string application_address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The address of the application this session is for
+    // TODO(@Olshansk): Uncomment the line below once the `ServiceId` proto is defined
+    // service.ServiceId service_id = 2; // The ID of the service this session is servicing
+    int64 session_block_start_height = 3; // The height at which this session started
```
Comment (bryanchriswhite) [inline proto/pocket/session/session.proto:20]: nit: maybe this would read slightly better as `session_start_block_height`?

**Comment by Olshansk** [comment]
@bryanchriswhite Going to merge this in and discuss the location of the type in your PR.

**Pull request merged**